dsgrid/query/models.py

```diff
@@ -0,0 +1,305 @@
+        allowed = ["avg", "min", "mean", "max", "sum"]
+        if operation not in allowed:
+            raise ValueError(f"operation={operation} is not supported. Allowed={allowed}")
+        return operation
+
+
+class AggregationModel(DSGBaseModel):
+
+    group_by_columns: List[str] = Field(
+        title="group_by_columns", description="Columns on which to group"
+    )
+    aggregation_function: Any = Field(
+        title="aggregation_function",
+        description="Must be a function name in pyspark.sql.functions",
+    )
+    name: Optional[str]
+
+    def dict(self, *args, **kwargs):
+        data = super().dict(*args, **kwargs)
+        data["aggregation_function"] = data["aggregation_function"].__name__
+        return data
+
+    @validator("group_by_columns")
+    def check_group_by_columns(cls, group_by_columns):
+        if not group_by_columns:
+            raise ValueError("group_by_columns cannot be empty")
+        return group_by_columns
+
+    @validator("aggregation_function")
+    def check_aggregation_function(cls, aggregation_function):
+        if isinstance(aggregation_function, str):
+            aggregation_function = getattr(F, aggregation_function, None)
+            if aggregation_function is None:
+                raise ValueError(f"{aggregation_function} is not defined in pyspark.sql.functions")
+        return aggregation_function
+
+
+class ChainedAggregationModel(DSGBaseModel):
+    aggregations: List[AggregationModel]
+
+    @validator("aggregations")
+    def check_aggregations(cls, aggregations):
+        length = len(aggregations)
+        if len(aggregations) < 2:
+            raise ValueError(f"length of ChainedAggregationModel must be at least: {length}")
+        if aggregations[-1].name is None:
+            raise ValueError("ChainedAggregationModel requires its last model to define a name.")
+        return aggregations
+
+
+# TODO: Deserializing JSON into QueryModel probably won't work for all cases, notably the unions.
+
+
+class ProjectConstraintsModel(DSGBaseModel):
+    """Defines how to transform a project into a derived dataset"""
+
+    project_id: str = Field(title="project_id", description="Project ID for query")
+    dataset_ids: List[str] = Field(
+        title="dataset_ids", description="Dataset IDs from which to read"
+    )
+    excluded_dataset_ids: List[str] = Field(
+        title="excluded_dataset_ids", description="Datasets to exclude from query", default=[]
+    )
+    include_dsgrid_dataset_components: bool = Field(
+        title="include_dsgrid_dataset_components", description=""
+    )
+    metric_reductions: List[MetricReductionModel] = Field(
+        title="metric_reductions",
+        description="Specifies how metric values should be reduced.",
+        default=[],
+    )
+    drop_dimensions: List[DimensionType] = Field(
+        title="drop_dimensions",
+        description="Drop columns for these dimensions.",
+        default=[],
```

Comment:
I forget who asked for this. I think the idea was that the user doesn't want to see a particular dimension in the final table because they don't care about it. In that case, there is a bug because this may produce duplicate rows that are not being dropped.

Reply:
Will leave this unimplemented for now. Add to TODO list.